nat/service_ics.go

```diff
@@ -0,0 +1,224 @@
+// +build windows
+
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package nat
+
+import (
+	"net"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+
+	log "github.com/cihub/seelog"
+	"github.com/pkg/errors"
+)
+
+type serviceICS struct {
+	mu                 sync.Mutex
+	ifaces             map[string]RuleForwarding // list in internal interfaces with enabled internet connection sharing
+	remoteAccessStatus string
+}
+
+// Enable enables internet connection sharing for the public interface.
+func (nat *serviceICS) Enable() error {
+	status, err := powerShell("Get-Service RemoteAccess | foreach { $_.StartType }")
+	if err != nil {
+		return errors.Wrap(err, "failed to get RemoteAccess service startup type")
+	}
+	nat.remoteAccessStatus = string(status)
+
+	if _, err := powerShell("Set-Service -Name RemoteAccess -StartupType automatic"); err != nil {
+		return errors.Wrap(err, "failed to set RemoteAccess service startup type to automatic")
+	}
+	if _, err := powerShell("Start-Service -Name RemoteAccess"); err != nil {
+		return errors.Wrap(err, "failed to start RemoteAccess service")
+	}
+
+	ifaceName, err := getPublicInterfaceName()
+	if err != nil {
+		return errors.Wrap(err, "failed to get public interface name")
+	}
+
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.EnableSharing(0)`)
+
+	return errors.Wrap(err, "failed to enable internet connection sharing")
+}
+
+// Add enables internet connection sharing for the local interface.
+func (nat *serviceICS) Add(rule RuleForwarding) error {
+	// TODO firewall rule configuration should be added here for new connections.
+	ifaceName, err := getInterfaceBySubnet(rule.SourceAddress)
+	if err != nil {
+		return errors.Wrap(err, "failed to get public interface name")
+	}
+
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
```

Comment:
this whole super string seems like a perfect candidate for a template since we're using it twice.